File: 0474-ones-and-zeroes/0474-ones-and-zeroes.py
class Solution:
    def findMaxForm(self, strs: List[str], m: int, n: int) -> int:
        dp = [[0]*(n + 1) for _ in range(m + 1)]

        for str in strs:
            countOne = str.count('1')
            countZero = str.count('0')
            for i in range(m, countZero - 1, -1):
                for j in range(n, countOne - 1, -1):
                    dp[i][j] = max(dp[i][j], dp[i - countZero][j - countOne] + 1)
        
        return dp[m][n]

File: 0474-ones-and-zeroes/0474-ones-and-zeroes.cpp
class Solution {
public:
    int findMaxForm(vector<string>& strs, int m, int n) {
        vector<vector<int>> dp(m + 1, vector<int>(n + 1, 0));

        for (string str : strs) {
            int countOne = 0;
            int countZero = 0;

            for (char c : str) {
                if (c == '0') {
                    countZero++;
                } else {
                    countOne++;
                }
            }

            for (int i = m; i >= countZero; i--) {
                for (int j = n; j >= countOne; j--) {
                    dp[i][j] = max(dp[i][j], dp[i - countZero][j - countOne] + 1);
                }
            }
        }

        return dp[m][n];
    }
};